File: src/main/java/io/github/boogiemonster1o1/cartses/item/ModItems.java
package io.github.boogiemonster1o1.cartses.item;

import java.lang.reflect.Field;

import io.github.boogiemonster1o1.cartses.Cartses;
import io.github.boogiemonster1o1.cartses.RegisterMe;
import io.github.boogiemonster1o1.cartses.entity.MinecartTypes;
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithBarrelEntity;
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithCraftingTableEntity;
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithEnderChestEntity;
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithGlowstoneEntity;
import io.github.boogiemonster1o1.cartses.entity.cart.MinecartWithRedstoneLampEntity;

import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.MinecartItem;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

public class ModItems {
	@RegisterMe("minecart_with_crafting_table")
	public static final Item MINECART_WITH_CRAFTING_TABLE = new MinecartItem(MinecartTypes.craftingTable(), new Item.Settings().group(ItemGroup.TRANSPORTATION));

	@RegisterMe("minecart_with_barrel")
	public static final Item MINECART_WITH_BARREL = new MinecartItem(MinecartTypes.barrel(), new Item.Settings().group(ItemGroup.TRANSPORTATION));

	@RegisterMe("minecart_with_glowstone")
	public static final Item MINECART_WITH_GLOWSTONE = new MinecartItem(MinecartTypes.glowstone(), new Item.Settings().group(ItemGroup.TRANSPORTATION));

	@RegisterMe("minecart_with_redstone_lamp")
	public static final Item MINECART_WITH_REDSTONE_LAMP = new MinecartItem(MinecartTypes.redstoneLamp(), new Item.Settings().group(ItemGroup.TRANSPORTATION));

	@RegisterMe("minecart_with_ender_chest")
	public static final Item MINECART_WITH_ENDER_CHEST = new MinecartItem(MinecartTypes.enderChest(), new Item.Settings().group(ItemGroup.TRANSPORTATION));

	public static void init() {
		for (Field field : ModItems.class.getFields()) {
			if (!RegisterMe.PREDICATE.test(field)) {
				continue;
			}
			RegisterMe annotation = field.getAnnotation(RegisterMe.class);
			try {
				Registry.register(Registry.ITEM, new Identifier(Cartses.MOD_ID, annotation.value()), (Item) field.get(null));
			} catch (IllegalAccessException e) {
				throw new AssertionError(e);
			}
		}
	}
}

File: src/main/scala/io/github/boogiemonster1o1/cartses/item/ModItems.scala
/*
 * Copyright (c) 2020 BoogieMonster1O1
 *
 * Permission to use, copy, modify, and/or distribute this software for
 * any purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
 * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

package io.github.boogiemonster1o1.cartses.item

import io.github.boogiemonster1o1.cartses.Cartses
import io.github.boogiemonster1o1.cartses.entity.MinecartTypes
import net.minecraft.item.{Item, ItemGroup, MinecartItem}
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

object ModItems {

	val minecartWithCraftingTable = new MinecartItem(MinecartTypes.craftingTable, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	val minecartWithBarrel = new MinecartItem(MinecartTypes.barrel, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	val minecartWithGlowstone = new MinecartItem(MinecartTypes.glowstone, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	val minecartWithRedstoneLamp = new MinecartItem(MinecartTypes.redstoneLamp, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	val minecartWithEnderChest = new MinecartItem(MinecartTypes.enderChest, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	val minecartWithNoteBlock = new MinecartItem(MinecartTypes.noteBlock, new Item.Settings().group(ItemGroup.TRANSPORTATION))

	def init(): Unit = {
		register("minecart_with_crafting_table", minecartWithCraftingTable)
		register("minecart_with_barrel", minecartWithBarrel)
		register("minecart_with_glowstone", minecartWithGlowstone)
		register("minecart_with_redstone_lamp", minecartWithRedstoneLamp)
		register("minecart_with_ender_chest", minecartWithEnderChest)
		register("minecart_with_note_block", minecartWithNoteBlock)
	}

	private def register(name: String, item: Item): Unit = {
		Registry.register(Registry.ITEM, new Identifier(Cartses.modId, name), item)
	}
}
